subsys/bluetooth/host/ecc.c

```diff
@@ -0,0 +1,202 @@
+		return 0;
+	}
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_P256_PUBLIC_KEY, NULL, NULL);
+	if (err) {
+		BT_ERR("Sending LE P256 Public Key command failed");
+		atomic_clear_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY);
+		pub_key_cb = NULL;
+		return err;
+	}
+
+	return 0;
+}
+
+const uint8_t *bt_pub_key_get(void)
+{
+	if (IS_ENABLED(CONFIG_BT_USE_DEBUG_KEYS) &&
+	    BT_CMD_TEST(bt_dev.supported_commands, 41, 2)) {
+		return debug_public_key;
+	}
+
+	if (atomic_test_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY)) {
+		return pub_key;
+	}
+
+	return NULL;
+}
+
+static int hci_generate_dhkey_v1(const uint8_t *remote_pk)
+{
+	struct bt_hci_cp_le_generate_dhkey *cp;
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_GENERATE_DHKEY, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	memcpy(cp->key, remote_pk, sizeof(cp->key));
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_GENERATE_DHKEY, buf, NULL);
+}
+
+static int hci_generate_dhkey_v2(const uint8_t *remote_pk, uint8_t key_type)
+{
+	struct bt_hci_cp_le_generate_dhkey_v2 *cp;
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_GENERATE_DHKEY_V2, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	memcpy(cp->key, remote_pk, sizeof(cp->key));
+	cp->key_type = key_type;
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_GENERATE_DHKEY_V2, buf, NULL);
+}
+
+int bt_dh_key_gen(const uint8_t remote_pk[64], bt_dh_key_cb_t cb)
+{
+	int err;
+
+	if (dh_key_cb == cb) {
+		return -EALREADY;
+	}
+
+	if (dh_key_cb || atomic_test_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY)) {
+		return -EBUSY;
+	}
+
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY)) {
+		return -EADDRNOTAVAIL;
+	}
+
+	dh_key_cb = cb;
+
+	if (IS_ENABLED(CONFIG_BT_USE_DEBUG_KEYS) &&
+	    BT_CMD_TEST(bt_dev.supported_commands, 41, 2)) {
+		err = hci_generate_dhkey_v2(remote_pk,
+					    BT_HCI_LE_KEY_TYPE_DEBUG);
+	} else {
+		err = hci_generate_dhkey_v1(remote_pk);
+	}
+
+	if (err) {
+		dh_key_cb = NULL;
+		BT_WARN("Failed to generate DHKey (err %d)", err);
+		return err;
+	}
+
+	return 0;
+}
+
+void bt_hci_evt_le_pkey_complete(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_p256_public_key_complete *evt = (void *)buf->data;
+	struct bt_pub_key_cb *cb;
+
+	BT_DBG("status: 0x%02x", evt->status);
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY);
+
+	if (!evt->status) {
+		memcpy(pub_key, evt->key, 64);
+		atomic_set_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY);
+	}
+
+	for (cb = pub_key_cb; cb; cb = cb->_next) {
+		cb->func(evt->status ? NULL : pub_key);
+	}
+
+	pub_key_cb = NULL;
+}
+
+void bt_hci_evt_le_dhkey_complete(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_generate_dhkey_complete *evt = (void *)buf->data;
+
+	BT_DBG("status: 0x%02x", evt->status);
+
+	if (dh_key_cb) {
+		bt_dh_key_cb_t cb = dh_key_cb;
+
+		dh_key_cb = NULL;
+		cb(evt->status ? NULL : evt->dhkey);
+	}
+}
```

Comment:
I'm all for these types of modularizations of hci_core.c. Nice, clean module :+1: